### PR TITLE
feat: signin unblock

### LINKED ISSUE #4036: Add support for signin authorization

### PR DESCRIPTION
**Pull request by shane-tomlinson**
fixes #4036

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
Ugh, sorry for the gigantic diff. @vbudhram, @seanmonstar, @vladikoff - r?

Going to try to put onto a dev box with @seanmonstar's PRs too.

**Comment by shane-tomlinson** [comment]
And now some functional tests fail. Still, let's start review to get feedback on code.

File: app/scripts/models/account.js [hunk 1]
```diff
@@ -889,6 +888,30 @@ define(function (require, exports, module) {
...
+    sendUnblockEmail () {
+      return this._fxaClient.sendUnblockEmail(
+        this.get('email')
+      );
+    },
+
+    /**
+     * Reject a login authorization code.
+     *
+     * @param {string} code
```
Comment (vladikoff) [inline app/scripts/models/account.js]: `unblockCode` ?

File: app/scripts/models/verification/report-sign-in.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * A model to hold report-sign-in data. Model has a `report` function
```
Comment (vladikoff) [inline app/scripts/models/verification/report-sign-in.js]: " Model has a `report` function" sorry where?

File: app/scripts/templates/report_sign_in.mustache [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+      <form novalidate>
+        <div class="button-row">
+          <button id="submit-btn" type="submit" class="disabled">{{#t}}Report activity{{/t}}</button>
+        </div>
+      </form>
+    </section>
+
+    <div class="links">
+      <a href="https://support.mozilla.org">{{#t}}Why is this happening?{{/t}}</a>
```
Comment (vladikoff) [inline app/scripts/templates/report_sign_in.mustache]: missing support link

Comment (vladikoff) [inline app/scripts/templates/report_sign_in.mustache]: we probably need metrics on this click.

File: app/scripts/templates/sign_in_unblock.mustache [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<div id="main-content" class="card">
+  {{#error}}
+    <header>
+      <h1 id="fxa-signin-unblock-header">{{#t}}Authorize this sign-in{{/t}}</h1>
```
Comment (vladikoff) [inline app/scripts/templates/sign_in_unblock.mustache]: this header is in both places for `#error`, can we just move it out of the condition?

File: app/scripts/templates/report_sign_in.mustache [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      <div class="success"></div>
+      <div class="error"></div>
+
+      <p>
+        {{#t}}You received an email about attempted access to your account. Would you like to report this activity as suspicious?{{/t}}
+      </p>
+
+      <form novalidate>
+        <div class="button-row">
+          <button id="submit-btn" type="submit" class="disabled">{{#t}}Report activity{{/t}}</button>
```
Comment (vladikoff) [inline app/scripts/templates/report_sign_in.mustache]: nit: do we need to add new metrics for this view into docs? not sure if we list `.submit` events

File: app/scripts/templates/sign_in_unblock.mustache [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+      <div class="button-row">
+        <button type="submit" class="use-logged-in">{{#t}}Continue{{/t}}</button>
+      </div>
+
+    </form>
+
+    <div class="links">
+      <a id="resend" class="delayed-fadein left" href="#">{{#t}}Didn't arrive? Resend{{/t}}</a>
+      <a href="https://support.mozilla.org" class="delayed-fadein right">{{#t}}Why is this happening?{{/t}}</a>
```
Comment (vladikoff) [inline app/scripts/templates/sign_in_unblock.mustache]: need to track this, + utm params

File: app/scripts/views/elements/unblock-code-input.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  const textInput = require('views/elements/text-input');
+
+  const element = Object.create(textInput);
+
+  element.match = function ($el) {
+    return $el.attr('type') === 'text' && $el.hasClass('unblock-code');
+  };
+
+  element.validate = function () {
+    const isRequired = typeof this.attr('required') !== 'undefined';
```
Comment (vladikoff) [inline app/scripts/views/elements/unblock-code-input.js]: this seems like a weird check, especially if we are using jquery. might be able to simplify

File: app/scripts/views/mixins/signin-mixin.js [hunk 8]
```diff
@@ -20,19 +20,21 @@ define(function (require, exports, module) {
...
+     * @param {Account} account - account being signed in to
      *     @param {String} account.sessionToken
      *     Session token from the account
      * @param {string} [password] - the user's password. Can be null if
      *  user is signing in with a sessionToken.
+     * @param {string} [unblockCode] - an unblock code.
      * @return {object} promise
      */
-    signIn: function (account, password) {
+    signIn (account, password, unblockCode) {
```
Comment (vladikoff) [inline app/scripts/views/mixins/signin-mixin.js]: not sure how i feel about `unblockCode` being 3rd arg. should this be 
` signIn (account, password, options) {` then `options.unblockCode`?

File: app/scripts/views/mixins/signin-mixin.js [hunk 9]
```diff
@@ -64,6 +67,21 @@ define(function (require, exports, module) {
           }
 
           return self.onSignInSuccess(account);
+        })
+        .fail((err) => {
+          function isUnblockableError(err) {
+            return err.verificationMethod === 'email-captcha';
```
Comment (vladikoff) [inline app/scripts/views/mixins/signin-mixin.js]: move `'email-captcha'` into a `const` into file above ?

File: tests/functional/sign_in.js [hunk 10]
```diff
@@ -183,7 +183,7 @@ define([
     },
 
     'data-flow-begin attribute is set': function () {
-      this.remote
+      return this.remote
```
Comment (shane-tomlinson) [inline tests/functional/sign_in.js]: Revert this change - it'll be fixed by fixing #4219

File: app/scripts/views/mixins/signin-mixin.js [hunk 11]
```diff
@@ -64,6 +67,21 @@ define(function (require, exports, module) {
...
+          }
+
+          if (isUnblockableError(err)) {
```
Comment (vladikoff) [inline app/scripts/views/mixins/signin-mixin.js]: might need metrics events for the unblock err dance here

File: tests/functional/sign_in_blocked.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+        .then(function () {
+          return this.parent
+            .then(type('#unblock_code', unblockCode));
+        })
+        .then(click('button[type=submit]'))
+
+        .then(testElementExists('#fxa-settings-header'));
+    },
+
+    'report signin link uid broken': function () {
```
Comment (shane-tomlinson) [inline tests/functional/sign_in_blocked.js]: This test fails locally.

File: app/scripts/views/report_sign_in.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        uid: verificationInfo.get('uid')
+      });
+      const unblockCode = verificationInfo.get('unblockCode');
+
+      return this.user.rejectAccountUnblockCode(account, unblockCode)
+        .then(() => this.navigate('signin_reported'));
+    },
+
+    context () {
+      const verificationInfo = this._verificationInfo;
```
Comment (vladikoff) [inline app/scripts/views/report_sign_in.js]: do we need this `const` here, we only use it once below

File: app/scripts/views/report_sign_in.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define(function (require, exports, module) {
```
Comment (vladikoff) [inline app/scripts/views/report_sign_in.js]: nit: maybe a doc above about this view

File: app/scripts/views/sign_in_unblock.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    submit () {
+      const account = this.getAccount();
+      const password = this.model.get('password');
+      const unblockCode = this.getElementValue('#unblock_code');
+
+      return this.signIn(account, password, unblockCode)
+        .fail((err) => this.onSignInError(account, password, err));
+    },
+
+    onSignInError (account, password, err) {
```
Comment (vladikoff) [inline app/scripts/views/sign_in_unblock.js]: hm surprised to see this error handler in `sign_in_unblock`.

File: app/scripts/models/verification/report-sign-in.js (hunk 2 above)
Comment (shane-tomlinson) [inline app/scripts/models/verification/report-sign-in.js]: Ahha, I took that out and moved it to `user`, good eye.

File: app/scripts/templates/report_sign_in.mustache (hunk 5 above)
Comment (shane-tomlinson) [inline app/scripts/templates/report_sign_in.mustache]: for sure we should add it.

File: app/scripts/templates/report_sign_in.mustache (hunk 3 above)
Comment (shane-tomlinson) [inline app/scripts/templates/report_sign_in.mustache]: And we need a real link. :D

File: app/scripts/templates/sign_in_unblock.mustache (hunk 4 above)
Comment (shane-tomlinson) [inline app/scripts/templates/sign_in_unblock.mustache]: Good point, I'll change this to something like "problem sending unblock email" because that's how this error condition occurs.

File: app/scripts/views/mixins/signin-mixin.js (hunk 9 above)
Comment (shane-tomlinson) [inline app/scripts/views/mixins/signin-mixin.js]: This should go into [lib/verification-methods.js](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/lib/verification-methods.js).

File: app/scripts/views/mixins/signin-mixin.js (hunk 11 above)
Comment (shane-tomlinson) [inline app/scripts/views/mixins/signin-mixin.js]: > might need metrics events for the unblock err dance here

@vladikoff  - is `fxa.content.screen.signin_unblock` sufficient?

File: app/scripts/views/sign_in_unblock.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define(function (require, exports, module) {
```
Comment (shane-tomlinson) [inline app/scripts/views/sign_in_unblock.js]: Add a top level doc.

File: app/scripts/views/sign_in_unblock.js (hunk 15 above)
Comment (shane-tomlinson) [inline app/scripts/views/sign_in_unblock.js]: We are re-signing the user in with the email/password they entered on the `/signin` page + the `unblockCode` entered here. If the user entered their password incorrectly, we'll only find out after this step, and then we have to send them back to the previous screen.

File: app/scripts/views/sign_in_unblock.js [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+      const unblockCode = this.getElementValue('#unblock_code');
+
+      return this.signIn(account, password, unblockCode)
+        .fail((err) => this.onSignInError(account, password, err));
+    },
+
+    onSignInError (account, password, err) {
+      if (AuthErrors.is(err, 'INCORRECT_PASSWORD')) {
+        // The user must go enter the correct password this time.
+        this.navigate('signin', {
```
Comment (shane-tomlinson) [inline app/scripts/views/sign_in_unblock.js]: Don't forget about `/oauth/signin` and users who arrive here from `/force_auth` and `/oauth/force_auth`.

File: tests/functional/sign_in_blocked.js (hunk 12 above)
Comment (shane-tomlinson) [inline tests/functional/sign_in_blocked.js]: Fixed it!

**Comment by vladikoff** [comment]
got a bit stuck here 
image

**Comment by vladikoff** [comment]
There is an issue with

> Correct Code ->> Incorrect Password ->> Press back button ->> arrive to "signin_unblock" with no query params


and I think that threw "invalid code". I'm guessing we would need to handle the "back" case somehow.

**Comment by shane-tomlinson** [comment]
> got a bit stuck here 

Hard blocks still remain in place to prevent signin unblock from being used as 1) a spam vector, and 2) a free pass to guess the user's FxA password by continually going through signin unblock. Some restrictions will likely be loosened a bit, @rfk wrote this in an email thread with @seanmonstar and I:

> I was able to get into a non-unblockable situation by doing the following:
>
> * Attempting signin to a bunch of non-existing accounts, then
> * Attempting signin with the wrong password on my real account
> 
> The result was a 429 while calling /send_unblock_code, and UX as shown
in the attached screenshot.
> 
> My mental model of what happened is that my IP was rate-limited due to
> the failed logins on different accounts, and since rate-limited IPs are
> not allowed to do anything, this blocked the sending of the
> authorization code email.
> 
> Sean, you might remember this failure mode from our time digging into
> that weird "Don't block email-sending on IP address alone" comment in
> the customs-server:
> 
> 
> https://github.com/mozilla/fxa-customs-server/commit/9a4eaf50f404656eb82e975a04a71e85437910e6
> 
> Perhaps the right thing in the short term, is to improve the UX of this
failure case, by only transitioning to the "we sent you an authorization
code" screen after we've successfully sent the code.

**Comment by shane-tomlinson** [comment]
> There is an issue with
>
> Correct Code ->> Incorrect Password ->> Press back button ->> arrive to "signin_unblock" with no query params
> 
> and I think that threw "invalid code". I'm guessing we would need to handle the "back" case somehow.

Good find!

File: app/scripts/views/elements/unblock-code-input.js (hunk 7 above)
Comment (shane-tomlinson) [inline app/scripts/views/elements/unblock-code-input.js]: We could create a plugin, the jquery docs say this:

> As of jQuery 1.6, the .attr() method returns undefined for attributes that have not been set.

File: app/scripts/models/user.js [hunk 18]
```diff
@@ -626,6 +627,17 @@ define(function (require, exports, module) {
           }
           return exists;
         });
+    },
+
+    /**
+     * Report/reject the unblockCode for the given account
```
Comment (vbudhram) [inline app/scripts/models/user.js]: This was a little confusing to me. I read it as either report or reject. Should it be reject and report? From looking at the [auth-server endpoint](https://github.com/mozilla/fxa-auth-server/blob/2c0561df2aae7d166a21c05b70aaaff6efe982e8/lib/routes/account.js#L1679), it appears to log an error, is this what is meant by report?

File: app/scripts/views/elements/unblock-code-input.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    const value = this.val();
+
+    if (isRequired && ! value.length) {
+      throw AuthErrors.toError('UNBLOCK_CODE_REQUIRED');
+    } else if (value.length !== Constants.UNBLOCK_CODE_LENGTH) {
```
Comment (vbudhram) [inline app/scripts/views/elements/unblock-code-input.js]: Can probably be replace with regex as well.

Comment (shane-tomlinson) [inline app/scripts/views/elements/unblock-code-input.js]: > Can probably be replace with regex as well.

I replaced this check with `Validate.isUnblockCodeValid`

**Comment by shane-tomlinson** [comment]
Problems/follow on work @ryanfeeley and I found today when testing:

* [ ] Incorrect password from force_auth does not allow the user to enter the correct password, upon return to /force_auth, the screen is blank except for an error that says "Incorrect password"
* [ ] Incorrect password from signup sends the user back to the signup page from signin_unblock, it should instead send the user to the signin page.
* [ ] Check how the screen appears from the firstrun page.
* [ ] Signin with an unverified account - the user must first enter the unblock code, THEN they see the "verify your account" screen.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 87119f8642e520e6586cb1d773b9f2916a21acdb
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 87119f8642e520e6586cb1d773b9f2916a21acdb
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 87119f8642e520e6586cb1d773b9f2916a21acdb
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/models/user.js (hunk 18 above)
Comment (shane-tomlinson) [inline app/scripts/models/user.js]: Yup.

File: app/scripts/views/force_auth.js [hunk 20]
```diff
@@ -57,6 +57,8 @@ define(function (require, exports, module) {
       try {
         accountData = this._getAndValidateAccountData();
       } catch (err) {
+        // validation errors are not handled here, rather they are
+        // handled XXX.
```
Comment (shane-tomlinson) [inline app/scripts/views/force_auth.js]: This comment should be expanded upon.

**Comment by shane-tomlinson** [comment]
> Signin with an unverified account - the user must first enter the unblock code, THEN they see the "verify your account" screen.

We have decided to punt on this for now.

File: app/tests/lib/helpers.js [hunk 21]
```diff
@@ -73,19 +73,20 @@ define(function (require, exports, module) {
...
-
-    for (var i = 0; i < length; ++i) {
-      var indexToUse = Math.floor(Math.random() * numberOfPossibilities);
-      str += lettersToChooseFrom.charAt(indexToUse);
+  function createRandomString(length, base = 36) {
+    let str = '';
+    let alphabet = '0123456789abcdefghijklmnopqrstuvwxyz';
+
+    for (let i = 0; i < length; ++i) {
+      let indexToUse = Math.floor(Math.random() * base);
```
Comment (philbooth) [inline app/tests/lib/helpers.js]: I know this is only test code, but it sounded yesterday like we were agreeing either to move `let` declarations outside of loops or to use `var` instead.

Comment (shane-tomlinson) [inline app/tests/lib/helpers.js]: I'm pretty "meh" since our for loops iterate over such a small range and that was a v8 specific issue.

Comment (shane-tomlinson) [inline app/tests/lib/helpers.js]: on server side code though, we should probably absolutely worry about it.

Comment (shane-tomlinson) [inline app/tests/lib/helpers.js]: probably absolutely. Yes. English, not my strongest subject.

File: app/scripts/models/user.js (hunk 18 above)
Comment (shane-tomlinson) [inline app/scripts/models/user.js]: Comment updated.

File: app/scripts/templates/report_sign_in.mustache (hunk 3 above)
Comment (shane-tomlinson) [inline app/scripts/templates/report_sign_in.mustache]: Remove the link until it the content is ready, or perhaps we can use an able experiment to insert the link when we have one ready.

File: app/scripts/views/sign_in_unblock.js [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    _sendUnblockEmail () {
+      return this.getAccount().sendUnblockEmail()
+        .fail((err) => this.displayError(err));
+    },
+
+    _getAuthPage () {
+      const lastPage = this.model.get('lastPage');
+
+      if (lastPage === 'force_auth') {
+        return 'force_auth';
```
Comment (shane-tomlinson) [inline app/scripts/views/sign_in_unblock.js]: Normalize these to handle the `oauth` case - this can be done using [this.broker.transformLink](https://github.com/mozilla/fxa-content-server/blob/9887e9ab0fa7a6cbaa30787733e0ac2c5802b00b/app/scripts/models/auth_brokers/base.js#L296:L304).

File: app/tests/spec/models/user.js [hunk 23]
```diff
@@ -1167,5 +1167,22 @@ define(function (require, exports, module) {
         assert.isFalse(exists);
       });
     });
+
+    describe('rejectAccountUnblockCode', () => {
+      let account;
+      let UNBLOCK_CODE = '12345678';
```
Comment (shane-tomlinson) [inline app/tests/spec/models/user.js]: This could be a `const`

**Comment by rfk** [comment]
> We have decided to punt on this for now.

@shane-tomlinson @seanmonstar please ensure we have a follow-up bug on file in the right milestone, so we don't lose track of it

File: app/tests/lib/helpers.js (hunk 21 above)
Comment (shane-tomlinson) [inline app/tests/lib/helpers.js]: > Why move the let outside the for? It seems like exactly what is designed for: block level scoping.

@jrgm thought he heard v8 has an issue where `let` inside of loops is significantly slower than `var`.

File: tests/functional/lib/helpers.js [hunk 24]
```diff
@@ -891,7 +934,7 @@ define([
         }, [command])
         .then(null, function (err) {
           if (/ScriptTimeout/.test(String(err))) {
-            var noSuchNotificationError = new Error('NoSuchBrowserNotification');
+            var noSuchNotificationError = new Error('NoSuchBrowserNotification ' + command);
```
Comment (shane-tomlinson) [inline tests/functional/lib/helpers.js]: revert this change.
